// @storybook/addon-comments/dist/manager/components/CommentList/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _style = require('./style');

var _style2 = _interopRequireDefault(_style);

var _CommentItem = require('../CommentItem');

var _CommentItem2 = _interopRequireDefault(_CommentItem);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var CommentList = function (_Component) {
  (0, _inherits3.default)(CommentList, _Component);

  function CommentList() {
    (0, _classCallCheck3.default)(this, CommentList);
    return (0, _possibleConstructorReturn3.default)(this, (CommentList.__proto__ || (0, _getPrototypeOf2.default)(CommentList)).apply(this, arguments));
  }

  (0, _createClass3.default)(CommentList, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      this.wrapper.scrollTop = this.wrapper.scrollHeight;
    }
  }, {
    key: 'componentDidUpdate',
    value: function componentDidUpdate(prev) {
      if (this.props.comments.length !== prev.comments.length) {
        this.wrapper.scrollTop = this.wrapper.scrollHeight;
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      var comments = this.props.comments;


      if (comments.length === 0) {
        return _react2.default.createElement(
          'div',
          {
            ref: function ref(el) {
              _this2.wrapper = el;
            },
            style: _style2.default.wrapper
          },
          _react2.default.createElement(
            'div',
            { style: _style2.default.noComments },
            'No Comments Yet!'
          )
        );
      }

      return _react2.default.createElement(
        'div',
        {
          ref: function ref(el) {
            _this2.wrapper = el;
          },
          style: _style2.default.wrapper
        },
        comments.map(function (comment) {
          return _react2.default.createElement(_CommentItem2.default, {
            key: comment.id,
            comment: comment,
            ownComment: comment.userId === (_this2.props.user && _this2.props.user.id),
            deleteComment: function deleteComment() {
              return _this2.props.deleteComment(comment.id);
            }
          });
        })
      );
    }
  }]);
  return CommentList;
}(_react.Component);

exports.default = CommentList;


CommentList.defaultProps = {
  comments: [],
  user: null,
  deleteComment: function deleteComment() {}
};

CommentList.propTypes = {
  comments: _propTypes2.default.arrayOf(_propTypes2.default.object),
  user: _propTypes2.default.shape({
    id: _propTypes2.default.number
  }),
  deleteComment: _propTypes2.default.func
};
