// @storybook/addon-knobs/dist/vue/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
var vueHandler = exports.vueHandler = function vueHandler(channel, knobStore) {
  return function (getStory) {
    return function (context) {
      return {
        data: function data() {
          return {
            context: context,
            getStory: getStory,
            story: getStory(context)
          };
        },
        render: function render(h) {
          return h(this.story);
        },


        methods: {
          onKnobChange: function onKnobChange(change) {
            var name = change.name,
                value = change.value;
            // Update the related knob and it's value.

            var knobOptions = knobStore.get(name);

            knobOptions.value = value;
            this.story = this.getStory(this.context);
            this.$forceUpdate();
          },
          onKnobClick: function onKnobClick(knob) {
            var knobOptions = knobStore.get(knob.name);
            knobOptions.callback();
          },
          onKnobReset: function onKnobReset() {
            knobStore.reset();
            this.setPaneKnobs(false);
            this.story = this.getStory(this.context);
            this.$forceUpdate();
          },
          setPaneKnobs: function setPaneKnobs() {
            var timestamp = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : +new Date();

            channel.emit('addon:knobs:setKnobs', { knobs: knobStore.getAll(), timestamp: timestamp });
          }
        },

        created: function created() {
          channel.on('addon:knobs:reset', this.onKnobReset);
          channel.on('addon:knobs:knobChange', this.onKnobChange);
          channel.on('addon:knobs:knobClick', this.onKnobClick);
          knobStore.subscribe(this.setPaneKnobs);
        },
        beforeDestroy: function beforeDestroy() {
          channel.removeListener('addon:knobs:reset', this.onKnobReset);
          channel.removeListener('addon:knobs:knobChange', this.onKnobChange);
          channel.removeListener('addon:knobs:knobClick', this.onKnobClick);
          knobStore.unsubscribe(this.setPaneKnobs);
        }
      };
    };
  };
};
