// @storybook/addon-comments/dist/manager/components/CommentItem/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _global = require('global');

var _moment = require('moment');

var _moment2 = _interopRequireDefault(_moment);

var _reactRenderHtml = require('react-render-html');

var _reactRenderHtml2 = _interopRequireDefault(_reactRenderHtml);

var _insertCss = require('insert-css');

var _insertCss2 = _interopRequireDefault(_insertCss);

var _style = require('./style');

var _style2 = _interopRequireDefault(_style);

var _style3 = require('./style.css');

var _style4 = _interopRequireDefault(_style3);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

(0, _insertCss2.default)(_style4.default);

var CommentItem = function (_Component) {
  (0, _inherits3.default)(CommentItem, _Component);

  function CommentItem(props) {
    var _ref;

    (0, _classCallCheck3.default)(this, CommentItem);

    for (var _len = arguments.length, args = Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
      args[_key - 1] = arguments[_key];
    }

    var _this = (0, _possibleConstructorReturn3.default)(this, (_ref = CommentItem.__proto__ || (0, _getPrototypeOf2.default)(CommentItem)).call.apply(_ref, [this, props].concat(args)));

    _this.state = { hovered: false };
    _this.mouseEnter = _this.mouseEnter.bind(_this);
    _this.mouseLeave = _this.mouseLeave.bind(_this);
    _this.deleteComment = _this.deleteComment.bind(_this);
    return _this;
  }

  (0, _createClass3.default)(CommentItem, [{
    key: 'mouseEnter',
    value: function mouseEnter() {
      this.setState({ hovered: true });
    }
  }, {
    key: 'mouseLeave',
    value: function mouseLeave() {
      this.setState({ hovered: false });
    }
  }, {
    key: 'deleteComment',
    value: function deleteComment() {
      var confirmDelete = _global.window.confirm('Are you sure you want to delete this comment?');
      if (confirmDelete === true) {
        this.props.deleteComment();
      }
    }
  }, {
    key: 'renderDelete',
    value: function renderDelete() {
      return _react2.default.createElement(
        'button',
        { type: 'button', style: _style2.default.commentDelete, onClick: this.deleteComment },
        'delete'
      );
    }
  }, {
    key: 'render',
    value: function render() {
      var comment = this.props.comment;

      var commentStyle = _style2.default.commentItem;
      if (comment.loading) {
        commentStyle = _style2.default.commentItemloading;
      }

      var time = (0, _moment2.default)(new Date(comment.time), 'YYYYMMDD').fromNow();
      var body = (0, _reactRenderHtml2.default)('<span>' + comment.text + '</span>');
      var showDelete = this.state.hovered && this.props.ownComment;

      return _react2.default.createElement(
        'div',
        { style: commentStyle, onMouseEnter: this.mouseEnter, onMouseLeave: this.mouseLeave },
        _react2.default.createElement(
          'div',
          { style: _style2.default.commentAside },
          _react2.default.createElement('img', { alt: comment.user.name, style: _style2.default.commentAvatar, src: comment.user.avatar })
        ),
        _react2.default.createElement(
          'div',
          { className: 'comment-content', style: _style2.default.commentContent },
          _react2.default.createElement(
            'div',
            { style: _style2.default.commentHead },
            _react2.default.createElement(
              'span',
              { style: _style2.default.commentUser },
              comment.user.name
            ),
            _react2.default.createElement(
              'span',
              { style: _style2.default.commentTime },
              time
            )
          ),
          _react2.default.createElement(
            'span',
            { style: _style2.default.commentText },
            body
          ),
          showDelete ? this.renderDelete() : null
        )
      );
    }
  }]);
  return CommentItem;
}(_react.Component);

exports.default = CommentItem;


CommentItem.defaultProps = {
  comment: {},
  deleteComment: function deleteComment() {},
  ownComment: false
};

CommentItem.propTypes = {
  deleteComment: _propTypes2.default.func,
  comment: _propTypes2.default.shape({
    user: _propTypes2.default.object,
    text: _propTypes2.default.string,
    time: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.instanceOf(Date)]),
    loading: _propTypes2.default.bool
  }),
  ownComment: _propTypes2.default.bool
};
