// @storybook/addon-comments/dist/manager/components/CommentForm/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _global = require('global');

var _reactTextareaAutosize = require('react-textarea-autosize');

var _reactTextareaAutosize2 = _interopRequireDefault(_reactTextareaAutosize);

var _marked = require('marked');

var _marked2 = _interopRequireDefault(_marked);

var _style = require('./style');

var _style2 = _interopRequireDefault(_style);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var renderer = new _marked2.default.Renderer();
renderer.heading = function (text) {
  return text;
};

_marked2.default.setOptions({
  renderer: renderer,
  gfm: true,
  tables: false,
  breaks: true,
  pedantic: false,
  sanitize: true,
  smartLists: true,
  smartypants: false
});

var CommentForm = function (_Component) {
  (0, _inherits3.default)(CommentForm, _Component);

  function CommentForm(props) {
    var _ref;

    (0, _classCallCheck3.default)(this, CommentForm);

    for (var _len = arguments.length, args = Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
      args[_key - 1] = arguments[_key];
    }

    var _this = (0, _possibleConstructorReturn3.default)(this, (_ref = CommentForm.__proto__ || (0, _getPrototypeOf2.default)(CommentForm)).call.apply(_ref, [this, props].concat(args)));

    _this.state = { text: '' };
    return _this;
  }

  (0, _createClass3.default)(CommentForm, [{
    key: 'onChange',
    value: function onChange(e) {
      var text = e.target.value;
      this.setState({ text: text });
    }
  }, {
    key: 'onSubmit',
    value: function onSubmit() {
      var addComment = this.props.addComment;

      var text = this.state.text.trim();
      if (!text || text === '') {
        return;
      }

      addComment((0, _marked2.default)(text));
      this.setState({ text: '' });
    }
  }, {
    key: 'openLogin',
    value: function openLogin() {
      var signInUrl = 'https://hub.getstorybook.io/sign-in?redirectUrl=' + encodeURIComponent(_global.window.location.href);
      _global.window.location.href = signInUrl;
    }
  }, {
    key: 'handleKeyDown',
    value: function handleKeyDown(e) {
      if (e.key === 'Enter' && !e.shiftKey) {
        e.preventDefault();
        this.onSubmit();
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      if (!this.props.user) {
        return _react2.default.createElement(
          'div',
          { style: _style2.default.wrapper },
          _react2.default.createElement(_reactTextareaAutosize2.default, { style: _style2.default.input, disabled: true }),
          _react2.default.createElement(
            'button',
            { style: _style2.default.submitButton, onClick: function onClick() {
                return _this2.openLogin();
              } },
            'Sign in with Storybook Hub'
          )
        );
      }

      var text = this.state.text;

      return _react2.default.createElement(
        'div',
        { style: _style2.default.wrapper },
        _react2.default.createElement(_reactTextareaAutosize2.default, {
          style: _style2.default.input,
          onChange: function onChange(e) {
            return _this2.onChange(e);
          },
          onKeyDown: function onKeyDown(e) {
            return _this2.handleKeyDown(e);
          },
          placeholder: 'Add your comment...',
          value: text
        }),
        _react2.default.createElement(
          'button',
          { style: _style2.default.submitButton, onClick: function onClick() {
              return _this2.onSubmit();
            } },
          'Submit'
        )
      );
    }
  }]);
  return CommentForm;
}(_react.Component);

exports.default = CommentForm;


CommentForm.defaultProps = {
  user: null,
  addComment: function addComment() {}
};
CommentForm.propTypes = {
  user: _propTypes2.default.object, // eslint-disable-line react/forbid-prop-types
  addComment: _propTypes2.default.func
};
